### PR TITLE
Implement allowNegateOption.

### PR DESCRIPTION
**Pull request by mshima**
# Pull Request

Fixes https://github.com/tj/commander.js/issues/1343

## Problem

- We are currently using commander with unknown option parsing using [meow](https://github.com/sindresorhus/meow) that allows `--no-` automatically.
- We implement ~25 sub commands each providing their options.
- Almost every option contains 3 states: false, true and default (saved value).
- Adding a negate option to every option is not an option:
  - help will be too big
  - not every option is added the same way, some are generated, some has to be coded.

## Solution

- Add allowNegateOption() to Options.
- Add implicitNegateOptions() to Command that will set allowNegateOption() at Options.
- Fallback parsing `--no-foo` to `--foo` with false value if no `--no-foo` option was found.

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -34,6 +34,13 @@ class Option {
     }
     this.description = description || '';
     this.defaultValue = undefined;
+    this.negateOption = false;
+  }
+
+  allowNegateOption(allowNegateOption) {
+    allowNegateOption = (allowNegateOption === undefined) || allowNegateOption;
+    this.negateOption = !this.negate && this.long && !this.required && allowNegateOption;
```
Comment (shadowspawn) [inline index.js:42]: I don't think the `required` test is appropriate. Commander allows `--no-bar` to be added when there is a `--bar <value>`.

**Comment by shadowspawn** [comment]
Some initial comments. I am not sure about this.

An implicit negate for a single option is interesting, but overlaps with the current implementation. The original implementation assumed there would be just one of positive or negative option, but #795 added the ability to have both work together.

Turning on implicit negate options for a whole command is quite intrusive. Every option will have to cope with the option value potentially being `false`.

The implicit negate should probably appear in the help description.

File: index.js (hunk 1 above)
Comment (mshima) [reply index.js]: Seems you are right.

**Comment by mshima** [comment]
Took your addOption from https://github.com/tj/commander.js/pull/1331.

**Comment by shadowspawn** [comment]
Somewhat related issues:

- #939 Triage: Problems with negating boolean options
- #795 fixes behavior of --no-* options

**Comment by shadowspawn** [comment]
There are two separate ideas in the description and original code for this PR:
- adding automatic support for `--no`
- having single option support for `--no` (rather than adding a separate `--no` option)

Both are interesting, but I would like more interest and use cases from the community before adding either. At the moment the use case is to match a specific existing implementation.

The single option support  for `--no` has a disadvantage that is overlaps with the existing support, so needs to offer some compelling value to justify supporting multiple approaches.

The code changes in `parseOptions` in this PR are reasonably tidy. To match historical usage, the custom option processing function is only called on option-arguments and not when the option is used in a boolean way. The existing handling passes in null as the value, and tests for that in the listener. To get similar behaviour with the added `--no` code the listener could perhaps check for a boolean value in addition to null, or instead of null.

There are major changes to the options coming inCommander v7 and the changes have now landed on the `release/7.x` branch. Have a look at what is possible and what is missing there.

**Pull request closed without merging**